### PR TITLE
IGNITE-22698 Sql. Extend test coverage for numeric comparison with literal wider than numeric type

### PR DESCRIPTION
**Pull request by zstan**
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java [hunk 1]
```diff
@@ -604,6 +609,27 @@ public RelDataType deriveType(SqlValidatorScope scope, SqlNode expr) {
...
     }
 
+    /** Check string contain only digit characters.
+     *
+     * @param strVal String representation.
+     * @param pos Current validator position.
+     *
+     * @throws CalciteException if unexpected input string found.
+     **/
+    static void checkStringContainDigitsOnly(@Nullable String strVal, SqlParserPos pos) throws CalciteException {
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java:619]: CalciteException is a runtime exception, there is no need to declare it.
If this method is only used by the typecoercsion it is better to move it there.

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java [hunk 2]
```diff
@@ -116,6 +118,143 @@ public void testUnicodeStrings() {
...
+
+        assertQuery(format("SELECT * FROM tbl WHERE v IN ('{}'::{} + 1, '0')", lessUpper, typeName)).returns(checkReturn).check();
+
+        assertQuery("SELECT * FROM tbl WHERE v IN (? + 1, '0')").withParam(lessUpper).returns(checkReturn).check();
+        assertQuery(format("SELECT * FROM tbl WHERE v NOT IN ('{}' - '1', '0')", strUpper)).returns(checkReturn).check();
+        assertQuery("SELECT * FROM tbl WHERE v NOT IN (? - '1', '0')").withParam(checkReturn).returns(checkReturn).check();
+
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-22697 Comparison on BIGINT cast brings overflow
+        // BigDecimal moreThanUpperBoundExact = new BigDecimal(strUpper).add(new BigDecimal(1));
+        // assertQuery(format("SELECT * FROM tbl WHERE v < {}::DECIMAL(19, 0)", moreThanUpperBoundExact)).returns(checkReturn).check();
```
Comment (lowka) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java:242]: Should be fixed by https://issues.apache.org/jira/browse/IGNITE-22519.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java [hunk 3]
```diff
@@ -103,6 +141,17 @@ private boolean doBinaryComparisonCoercion(SqlCallBinding binding) {
         RelDataType leftType = binding.getOperandType(0);
         RelDataType rightType = binding.getOperandType(1);
 
+        if (SqlTypeUtil.isIntType(leftType) && SqlTypeUtil.isCharacter(rightType)
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java:144]: Could you please write a comment explaining a problem does this code solves.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java (hunk 1 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java]: got it

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java (hunk 2 above)
Comment (zstan) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java:246]: fixed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java (hunk 3 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java]: remove it

**Review by lowka — APPROVED** [review]

**Pull request merged**